src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/RenameDriver.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core
+{
+    public class RenameDriver
+    {
+        private readonly RecordType _parameters;
+        private readonly SimpleResolver _resolver;
+
+        internal RenameDriver(RecordType parameters, DPath pathToRename, DName updatedName, SimpleResolver resolver)
+        {
+            var segments = new Queue<DName>(pathToRename.Segments());
+            Contracts.CheckParam(segments.Count > 0, nameof(parameters));
+
+            // After this point, _parameters should have at most one logical->display pair that can change in this conversion
+            _parameters = RenameFormulaTypeHelper(parameters, segments, updatedName) as RecordType;
+            _resolver = resolver;
+        }
+
+        /// <summary>
+        /// Applies rename operation to <paramref name="expressionText"/>.
+        /// </summary>
+        /// <param name="expressionText">Expression in which to rename the parameter field.</param>
+        /// <returns>Expression with rename applied.</returns>
+        public string ApplyRename(string expressionText)
+        {
+            return Engine.ConvertExpression(expressionText, _parameters, _resolver, true);
+        }
+
+        private static FormulaType RenameFormulaTypeHelper(AggregateType nestedType, Queue<DName> segments, DName updatedName)
+        {
+            var field = segments.Dequeue();
+            if (segments.Count == 0)
```

Comment:
Are we assuming that the queue length will be either 0 or 1, shouldn't we have a parameter check for that assumption?

Reply:
Wait, I see this is recursive